core/src/impl/Kokkos_ViewHookDeepCopy.hpp

```diff
@@ -0,0 +1,134 @@
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_VIEWHOOK_DEEPCOPY_HPP
+#define KOKKOS_VIEWHOOK_DEEPCOPY_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+/**
+  * Here we define copy_buffer utility functions
+  * to copy data to and from Views. We need
+  * specialized variants for ScratchMemorySpace
+  * and AnonymousSpace as they don't participate
+  * in the ViewHooks system, and we need copies
+  * to and from Views
+  */
+template <class ViewType>
+typename std::enable_if<(std::is_same<Kokkos::AnonymousSpace,
+                                      typename ViewType::memory_space>::value ||
+                         std::is_same<Kokkos::ScratchMemorySpace<
+                                          typename ViewType::execution_space>,
+                                      typename ViewType::memory_space>::value),
+                        void>::type
+copy_buffer(unsigned char *, ViewType &) {}
+
+template <class ViewType>
+typename std::enable_if<
+    !(std::is_same<Kokkos::AnonymousSpace,
+                   typename ViewType::memory_space>::value ||
+      std::is_same<
+          Kokkos::ScratchMemorySpace<typename ViewType::execution_space>,
+          typename ViewType::memory_space>::value),
+    void>::type
+copy_buffer(unsigned char *buff, ViewType &v) {
+  Kokkos::Impl::DeepCopy<Kokkos::HostSpace, typename ViewType::memory_space,
+                         typename ViewType::execution_space>(
+      (void *)buff, v.data(), v.span() * sizeof(typename ViewType::value_type));
+}
+
+template <class ViewType>
+typename std::enable_if<(std::is_same<Kokkos::AnonymousSpace,
+                                      typename ViewType::memory_space>::value ||
+                         std::is_same<Kokkos::ScratchMemorySpace<
+                                          typename ViewType::execution_space>,
+                                      typename ViewType::memory_space>::value ||
```

Comment:
This constraint is used multiple times. Please name it (with a inline variable template or something similar) instead of copy-pasting.